src/cloud-api-adaptor/pkg/securecomms/ppssh/ppssh.go

```diff
@@ -0,0 +1,313 @@
+	}
+	return nil
+}
+
+func (s *SshServer) Start(ctx context.Context) error {
+	var err error
+
+	logger.Printf("SSH service starting on port: %s", s.sshport)
+	s.ctx = ctx
+	s.listener, err = net.Listen("tcp", "0.0.0.0:"+s.sshport)
+	if err != nil {
+		logger.Fatal("Failed to listen for connection: ", err)
+	}
+
+	go func() {
+		kubernetesPhaseConfig := s.attestationPhase()
+		if kubernetesPhaseConfig == nil {
+			logger.Fatal("Attestation phase failed")
+		}
+		for ctx.Err() == nil {
+			s.kubernetesPhase(kubernetesPhaseConfig)
+		}
+		s.listener.Close()
+	}()
+	return nil
+}
+
+func Singleton() {
+	// Singleton - make sure we run the ssh service once per boot.
+	if _, err := os.Stat(SINGLETON_PATH); !errors.Is(err, os.ErrNotExist) {
+		logger.Fatal("SSH service runs in singleton mode and cannot be executed twice")
+	}
+	singleton, err := os.Create(SINGLETON_PATH)
+	if err != nil {
+		logger.Fatalf("Failed to create Singleton file: %v", err)
+	}
+	singleton.Close()
+}
+
+func getAttestationPhaseKeys() (ppPrivateKeyBytes []byte, tePublicKeyBytes []byte) {
+	var err error
+
+	// Attestation phase - may have unproven tePublicKeyBytes
+	tePublicKeyBytes, err = os.ReadFile(UNPROVEN_WN_PUBLIC_KEY_PATH)
+	if err != nil {
+		tePublicKeyBytes = nil
+	}
+
+	// Private Key generation - unproven to the client, key is generated on the fly
+	ppPrivateKey, err := rsa.GenerateKey(rand.Reader, 3072)
+	if err != nil {
+		logger.Fatalf("Attestation phase: failed to generate host key, err: %v", err)
+	}
+
+	// Validate Private Key
+	err = ppPrivateKey.Validate()
+	if err != nil {
+		logger.Fatalf("Attestation phase: failed to validate host key, err: %v", err)
+	}
+
+	ppPrivateKeyBytes = sshutil.RsaPrivateKeyPEM(ppPrivateKey)
+	logger.Printf("Attestation phase: SSH server initialized keys")
+	return
+}
+
+func setConfigHostKey(config *ssh.ServerConfig, ppPrivateKeyBytes []byte) error {
+	serverSigner, err := ssh.ParsePrivateKey(ppPrivateKeyBytes)
+	if err != nil {
+		return fmt.Errorf("unable to parse private key: %w", err)
+	}
+	config.AddHostKey(serverSigner)
+	return nil
+}
+
+func setPublicKey(config *ssh.ServerConfig, tePublicKeyBytes []byte) error {
+	teSshPublicKey, _, _, _, err := ssh.ParseAuthorizedKey(tePublicKeyBytes)
+	if err != nil {
+		return fmt.Errorf("unable to parse public key: %w", err)
+	}
+
+	// An SSH server is represented by a ServerConfig, which holds
+	// certificate details and handles authentication of ServerConns.
+	config.PublicKeyCallback = func(c ssh.ConnMetadata, clientPublicKey ssh.PublicKey) (*ssh.Permissions, error) {
+		if bytes.Equal(teSshPublicKey.Marshal(), clientPublicKey.Marshal()) {
+			return &ssh.Permissions{
+				// Record the public key used for authentication.
+				Extensions: map[string]string{
+					"pubkey-fp": ssh.FingerprintSHA256(clientPublicKey),
+				},
+			}, nil
+		}
+		return nil, fmt.Errorf("unknown public key for %q", c.User())
+	}
+	return nil
+}
+
```

Comment:
any specific reason to make the following methods global? I don't see it used outside of this package..

Reply:
changed to not expose these methods